config/src/main/java/com/typesafe/config/impl/ConfigBeanImpl.java

```diff
@@ -90,12 +90,17 @@
                     Path path = Path.newKey(name);
                     AbstractConfigValue configValue = configProps.get(beanProp.getName());
                     if (configValue != null) {
-                        SimpleConfig.checkValid(path, expectedType, configValue, problems);
-                    } else {
-                        if (!isOptionalProperty(clazz, beanProp)) {
-                            SimpleConfig.addMissing(problems, expectedType, path, config.origin());
+                        if (ConfigValueType.NULL != configValue.valueType()) {
+                            SimpleConfig.checkValid(path, expectedType, configValue, problems);
+                        } else {
+                            // Treat null values as if they are missing
+                            originalNames.remove(beanProp.getName());
+                            configValue = null;
                         }
                     }
+                    if (configValue == null && !isOptionalProperty(clazz, beanProp)) {
+                        SimpleConfig.addMissing(problems, expectedType, path, config.origin());
+                    }
```

Comment:
Does this imply that the `throw new ConfigException.Missing` a few lines down is now unreachable? If so we might throw a `BugOrBrokenException("Should have detected missing props earlier")` down there instead of `ConfigException.Missing`.